config/v1/types_infrastructure.go

```diff
@@ -1277,6 +1310,86 @@ type VSpherePlatformTopology struct {
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup.vmHostRule) : !has(self.hostGroup) || !has(self.hostGroup.vmHostRule)",message="vmHostRule is required when type is HostGroup, and forbidden otherwise"
+// +union
+type VSphereFailureDomainZoneAffinity struct {
+	// type is the string representation of the VSphereFailureDomainZoneType with available options of
+	// ComputeCluster and HostGroup.
+	// +kubebuilder:validation:Enum:=HostGroup;ComputeCluster
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Type:=string
+	// +unionDiscriminator
+	Type VSphereFailureDomainZoneType `json:"type"`
+
+	// hostGroup holds the vmGroup and the hostGroup names in vCenter
+	// corresponds to a vm-host group of type Virtual Machine and Host respectively. Is also
+	// contains the vmHostRule which is an affinity vm-host rule in vCenter.
+	// +unionMember,optional
+	HostGroup *VSphereFailureDomainHostGroup `json:"hostGroup,omitempty"`
+	// computeCluster currently is empty for future use.
+	// +unionMember,optional
+	ComputeCluster *VSphereFailureDomainComputeCluster `json:"computeCluster,omitempty"`
+}
+
+// VSphereFailureDomainRegionAffinity contains the region type and two currently
+// empty fields, datacenter and computeCluster for future use.
+type VSphereFailureDomainRegionAffinity struct {
+	// type is the string representation of the VSphereFailureDomainRegionType with available options of
+	// Datacenter and ComputeCluster
+	// +kubebuilder:validation:MinLength=9
+	// +kubebuilder:validation:MaxLength=14
+	// +kubebuilder:validation:Enum:=ComputeCluster;Datacenter
+	// +kubebuilder:validation:Type:=string
+	// +unionDiscriminator
+	Type VSphereFailureDomainRegionType `json:"type"`
+
+	// datacenter currently is empty for future use.
+	// +unionMember,optional
+	Datacenter *VSphereFailureDomainDatacenter `json:"datacenter,omitempty"`
+	// computeCluster currently is empty for future use.
+	// +unionMember,optional
+	ComputeCluster *VSphereFailureDomainComputeCluster `json:"computeCluster,omitempty"`
+}
+
+// VSphereFailureDomainHostGroup holds the vmGroup and the hostGroup names in vCenter
+// corresponds to a vm-host group of type Virtual Machine and Host respectively. Is also
+// contains the vmHostRule which is an affinity vm-host rule in vCenter.
+type VSphereFailureDomainHostGroup struct {
+	// vmGroup is the name of the vm-host group of type virtual machine within vCenter for this failure domain.
+	// vmGroup is limited to 80 characters.
+	// This field is required when the VSphereFailureDomain ZoneType is HostGroup
+	// +kubebuilder:validation:MaxLength=80
+	// +kubebuilder:validation:Optional
```

Comment:
I believe all three of these should be required now right?

Reply:
Using the current CEL and required causes the unit tests to fail. based on the error required is checked first. 

Unless the unit tests should be changed since it isn't possible to have those fields empty?

Reply:
Perhaps the CEL is wrong, I'd expect CEL to require the parent struct, and all fields within then to be required, but not necessarily have CEL within them? 

Required strings probably want a minLength as well right?